File: examples/rdma/python/py_rdma/s3rdmaput.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""s3rdmaput.py - upload a file to S3 via RDMA, using rdmalib + s3client.
Python 2.7 / 3.x compatible.

Usage:
  python s3rdmaput.py -bucket <bucket> -object <key> -file <path>
                       [-endpoint http://127.0.0.1:8901] [-rdma-dev rxe0]
                       [-ak <access-key>] [-sk <secret-key>]
"""
from __future__ import print_function

import argparse
import os
import sys

try:                       # Python 3
    from urllib.parse import urlsplit
except ImportError:        # Python 2
    from urlparse import urlsplit

import rdmalib
import s3client


def ws_url_from_endpoint(endpoint):
    parts = urlsplit(endpoint)
    return "ws://%s/rdma-ctrl" % (parts.netloc,)


def main():
    p = argparse.ArgumentParser()
    p.add_argument("-bucket", required=True, help="Name of the bucket")
    p.add_argument("-object", required=True, help="Name of the object")
    p.add_argument("-file", required=True, help="Path to the file to upload")
    p.add_argument("-endpoint", default="http://127.0.0.1:8901", help="S3 endpoint URL")
    p.add_argument("-rdma-dev", dest="rdma_dev", default="rxe0", help="RDMA device name")
    p.add_argument("-ak", default="admin-ak", help="S3 access key")
    p.add_argument("-sk", default="admin-sk", help="S3 secret key")
    args = p.parse_args()

    cfg = s3client.S3Config(endpoint=args.endpoint, region="us-east-1",
                             access_key=args.ak, secret_key=args.sk)

    # Step 1: read the file into memory
    try:
        with open(args.file, "rb") as f:
            data = f.read()
    except IOError as e:
        print("Failed to read file: %s" % e, file=sys.stderr)
        return 1

    request_id = "rdma-%d" % os.getpid()

    # Step 2: RDMA handshake (register, QP setup, copy data into mmap'd MR, send token)
    ws_url = ws_url_from_endpoint(args.endpoint)
    print("Initializing RDMA resources and exchanging QP info...", file=sys.stderr)
    try:
        sess = rdmalib.RDMASession.send_setup(ws_url, args.rdma_dev, rdmalib.RDMA_PORT,
                                               rdmalib.RDMA_GID_IDX, request_id, data)
    except Exception as e:
        print("RDMA setup failed: %s" % e, file=sys.stderr)
        return 1

    # Step 3: trigger the server-side RDMA read via S3 PutObject (empty body)
    print("Sending S3 PutObject request...", file=sys.stderr)
    try:
        s3client.put_object(cfg, args.bucket, args.object,
                             ("X-RDMA-Request-ID", request_id))
    except Exception as e:
        print("PutObject error: %s" % e, file=sys.stderr)
        sess.destroy()
        return 1

    print("Successfully uploaded %s to s3://%s/%s via RDMA (HTTP 200 OK)" %
          (args.file, args.bucket, args.object), file=sys.stderr)

    # Step 4: RDMA read is complete now; release RDMA resources
    sess.destroy()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: examples/rdma/java/src/main/java/com/example/s3rdma/S3RdmaPut.java
package com.example.s3rdma;

import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Paths;

/**
 * s3rdmaput - upload a file to S3 via RDMA, using RdmaSession + S3Client.
 *
 * Usage:
 *   java -Djna.library.path=native -cp target/s3rdma.jar com.example.s3rdma.S3RdmaPut \
 *        -bucket <bucket> -object <key> -file <path>
 *        [-endpoint http://127.0.0.1:8901] [-rdma-dev rxe0] [-ak <ak>] [-sk <sk>]
 */
public class S3RdmaPut {

    private static void usage() {
        System.err.println("Usage: S3RdmaPut -bucket <bucket> -object <key> -file <path> "
                + "[-endpoint http://127.0.0.1:8901] [-rdma-dev rxe0] [-ak <ak>] [-sk <sk>]");
        System.exit(1);
    }

    private static String wsUrlFromEndpoint(String endpoint) throws Exception {
        URI u = URI.create(endpoint);
        String authority = u.getAuthority();
        return "ws://" + authority + "/rdma-ctrl";
    }

    public static void main(String[] args) throws Exception {
        String bucket = null, object = null, filePath = null;
        String endpoint = "http://127.0.0.1:8901";
        String rdmaDev = "rxe0";
        String ak = "admin-ak", sk = "admin-sk";

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "-bucket": bucket = args[++i]; break;
                case "-object": object = args[++i]; break;
                case "-file": filePath = args[++i]; break;
                case "-endpoint": endpoint = args[++i]; break;
                case "-rdma-dev": rdmaDev = args[++i]; break;
                case "-ak": ak = args[++i]; break;
                case "-sk": sk = args[++i]; break;
                default: usage();
            }
        }
        if (bucket == null || object == null || filePath == null) {
            usage();
        }

        S3Client cfg = new S3Client(endpoint, "us-east-1", ak, sk);

        // Step 1: read the file into memory
        byte[] data;
        try {
            data = Files.readAllBytes(Paths.get(filePath));
        } catch (Exception e) {
            System.err.println("Failed to read file: " + e.getMessage());
            System.exit(1);
            return;
        }

        String requestId = "rdma-" + ProcessHandle.current().pid();

        // Step 2: RDMA handshake (register, QP setup, copy data into MR, send token)
        String wsUrl = wsUrlFromEndpoint(endpoint);
        System.err.println("Initializing RDMA resources and exchanging QP info...");
        RdmaSession sess;
        try {
            sess = RdmaSession.sendSetup(wsUrl, rdmaDev, RdmaSession.RDMA_PORT, RdmaSession.RDMA_GID_IDX,
                    requestId, data);
        } catch (Exception e) {
            System.err.println("RDMA setup failed: " + e.getMessage());
            System.exit(1);
            return;
        }

        // Step 3: trigger the server-side RDMA read via S3 PutObject (empty body)
        System.err.println("Sending S3 PutObject request...");
        try {
            cfg.putObject(bucket, object, "X-RDMA-Request-ID", requestId);
        } catch (Exception e) {
            System.err.println("PutObject error: " + e.getMessage());
            sess.destroy();
            System.exit(1);
            return;
        }

        System.err.println("Successfully uploaded " + filePath + " to s3://" + bucket + "/" + object
                + " via RDMA (HTTP 200 OK)");

        // Step 4: RDMA read is complete now; release RDMA resources
        sess.destroy();
    }
}
